src/lightmap.cpp

```diff
@@ -490,98 +536,126 @@ void map::build_seen_cache( const tripoint &origin, const int target_z )
+        castLight<-1, 0, 0, 1, sight_calc, sight_check>(
+            seen_cache, transparency_cache, origin.x, origin.y, 0 );
 
-    castLight<0, -1, -1, 0, sight_calc, sight_check>(
-        seen_cache, transparency_cache, origin.x, origin.y, 0 );
-    castLight<-1, 0, 0, -1, sight_calc, sight_check>(
-        seen_cache, transparency_cache, origin.x, origin.y, 0 );
+        castLight<0, 1, -1, 0, sight_calc, sight_check>(
+            seen_cache, transparency_cache, origin.x, origin.y, 0 );
+        castLight<1, 0, 0, -1, sight_calc, sight_check>(
+            seen_cache, transparency_cache, origin.x, origin.y, 0 );
 
-    int part;
-    if ( vehicle *veh = veh_at( origin, part ) ) {
-        // We're inside a vehicle. Do mirror calcs.
-        std::vector<int> mirrors = veh->all_parts_with_feature(VPFLAG_EXTENDS_VISION, true);
-        // Do all the sight checks first to prevent fake multiple reflection
-        // from happening due to mirrors becoming visible due to processing order.
-        // Cameras are also handled here, so that we only need to get through all veh parts once
-        int cam_control = -1;
-        for (std::vector<int>::iterator m_it = mirrors.begin(); m_it != mirrors.end(); /* noop */) {
-            const auto mirror_pos = veh->global_pos() + veh->parts[*m_it].precalc[0];
-            // We can utilize the current state of the seen cache to determine
-            // if the player can see the mirror from their position.
-            if( !veh->part_info( *m_it ).has_flag( "CAMERA" ) &&
-                seen_cache[mirror_pos.x][mirror_pos.y] < LIGHT_TRANSPARENCY_SOLID + 0.1 ) {
-                m_it = mirrors.erase(m_it);
-            } else if( !veh->part_info( *m_it ).has_flag( "CAMERA_CONTROL" ) ) {
-                ++m_it;
-            } else {
-                if( origin.x == mirror_pos.x && origin.y == mirror_pos.y && veh->camera_on ) {
-                    cam_control = *m_it;
-                }
-                m_it = mirrors.erase( m_it );
-            }
+        castLight<0, -1, -1, 0, sight_calc, sight_check>(
+            seen_cache, transparency_cache, origin.x, origin.y, 0 );
+        castLight<-1, 0, 0, -1, sight_calc, sight_check>(
+            seen_cache, transparency_cache, origin.x, origin.y, 0 );
+    } else {
+        if( origin.z == target_z ) {
+            seen_cache[origin.x][origin.y] = LIGHT_TRANSPARENCY_CLEAR;
         }
 
-        for( size_t i = 0; i < mirrors.size(); i++ ) {
-            const int &mirror = mirrors[i];
-            bool is_camera = veh->part_info( mirror ).has_flag( "CAMERA" );
-            if( is_camera && cam_control < 0 ) {
-                continue; // Player not at camera control, so cameras don't work
-            }
+        const auto floor_check = [this]( const tripoint &p ) {
+            const tripoint below( p.x, p.y, p.z - 1 );
+            return !valid_move( p, below, false, true );
+        };
+
+        // Raycasting, as opposed to shadowcasting for 2D
```

Comment:
I see what you mean.
I had a similar idea.

Well, I'm not sure if I get what minor and major axes are here, but otherwise I think I get it.
If minor and major axes are arbitrarily chosen (like now y vs x in `castLight`) rather than depending on some factor, then I'm pretty sure I get it.